### PR TITLE
Adds a JRuby optimized version of Concurrent::Atom

### PR DESCRIPTION
**Pull request by lucasallan**
Non-optimized version
```
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   4.720000   0.030000   4.750000 (  4.103692)
After JVM warm-up...
   3.920000   0.020000   3.940000 (  3.925786)
```
Optimized version
```
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   2.980000   0.020000   3.000000 (  2.742403)
After JVM warm-up...
   2.440000   0.010000   2.450000 (  2.479192)
```

Benchmark script:
```ruby
#!/usr/bin/env ruby

require 'concurrent'
require 'benchmark'

NUM      = 50_000_000
WARM_UPS = NUM

def jruby?
  defined? JRUBY_VERSION
end

def version
  gem_dir = Gem::Specification.find_by_name('concurrent-ruby').gem_dir
  File.open(File.join(gem_dir, 'VERSION'), 'r').read
rescue
  'unknown'
end

puts "~~~ Ruby version: #{RUBY_VERSION}"
if jruby?
  puts "~~~ JRuby version: #{JRUBY_VERSION}"
else
  puts "~~~ Ruby engine: #{RUBY_ENGINE}"
end
puts "~~~ Gem version: #{version}"

Benchmark.bm do |stats|

  puts "Benchmarking #{Concurrent::Atom}..."

  ref   = Concurrent::Atom.new('foo')
  value = nil

  puts 'Before JVM warm-up...' if jruby?
  stats.report do
    NUM.times { value = ref.value }
  end

  if jruby?
    WARM_UPS.times { value = ref.value }

    puts 'After JVM warm-up...'
    stats.report do
      NUM.times { value = ref.value }
    end
  end
end
```

### CODE REVIEW

File: lib/concurrent/jruby_atom.rb [hunk 1]
```diff
@@ -0,0 +1,223 @@
+if Concurrent.on_jruby?
+
+  require 'concurrent/collection/copy_on_notify_observer_set'
+  require 'concurrent/concern/observable'
+  require 'java'
+
+  java_import 'java.util.concurrent.atomic.AtomicReference'
```
Comment (ianks) [inline lib/concurrent/jruby_atom.rb]: Out of curiosity, how big is the perf hit of using Concurrent::Reference versus the Java one?

**Comment by jdantonio** [comment]
@lucasallan Thank you very much for working on this. I'm having difficulty following the differences between the implementations. AFAICT the main difference is that the newer version simply uses `java.util.concurrent.atomic.AtomicReference` directly rather than `Concurrent::AtomicReference`. Is that correct?

**Comment by lucasallan** [comment]
@jdantonio This version is directly based on Clojure code[1], but I guess we should be using `Concurrent::AtomicReference` rather than `java.util.concurrent.atomic.AtomicReference`.
But before it wasn't using `AtomicReference` anyway, right?

[1] - https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/Atom.java

**Comment by jdantonio** [comment]
I didn't look at the Clojure code before implementing this or Agent. I worked entirely from the documented behavior. I probably should have looked at the Clojure source. :-) If the implementation can be improved following that example I'm all for it, but can we improve it for the base implementation? Do we need a JRuby-specific version?

**Comment by lucasallan** [comment]
If we build an implementation around `AtomicReference` then we don't need a JRuby optimized version, since `AtomicReference` is already jruby-optimized.

**Comment by jdantonio** [comment]
@lucasallan Thank you!

**Comment by lucasallan** [comment]
@jdantonio Results - I executed it 4 times: 

### Using `Concurrent::Atomic` 
```
➜  concurrent-ruby git:(master) ✗ ruby ../atom_bench.rb
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   4.190000   0.040000   4.230000 (  3.116446)
After JVM warm-up...
   2.840000   0.010000   2.850000 (  2.848073)


➜  concurrent-ruby git:(master) ✗ ruby ../atom_bench.rb
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   2.860000   0.020000   2.880000 (  2.546445)
After JVM warm-up...
   2.400000   0.010000   2.410000 (  2.395978)



➜  concurrent-ruby git:(master) ✗ ruby ../atom_bench.rb
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   2.620000   0.020000   2.640000 (  2.438905)
After JVM warm-up...
   2.500000   0.010000   2.510000 (  2.504156)



➜  concurrent-ruby git:(master) ✗ ruby ../atom_bench.rb
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   2.740000   0.010000   2.750000 (  2.386496)
After JVM warm-up...
   2.390000   0.010000   2.400000 (  2.376449)

```

### Using concurrent-ruby-0.9.1(Current implementation):
```
➜  concurrent-ruby git:(master) ✗ ruby ../atom_bench.rb
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   4.690000   0.040000   4.730000 (  4.067498)
After JVM warm-up...
   3.690000   0.010000   3.700000 (  3.705187)


➜  concurrent-ruby git:(master) ✗ ruby ../atom_bench.rb
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   5.000000   0.030000   5.030000 (  4.547322)
After JVM warm-up...
   4.200000   0.010000   4.210000 (  4.217589)


➜  concurrent-ruby git:(master) ✗ ruby ../atom_bench.rb
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   4.710000   0.020000   4.730000 (  4.314722)
After JVM warm-up...
   4.290000   0.010000   4.300000 (  4.258959)


➜  concurrent-ruby git:(master) ✗ ruby ../atom_bench.rb
~~~ Ruby version: 2.2.2
~~~ JRuby version: 9.0.1.0
~~~ Gem version: unknown
       user     system      total        real
Benchmarking Concurrent::Atom...
Before JVM warm-up...
   4.320000   0.020000   4.340000 (  3.944181)
After JVM warm-up...
   3.820000   0.020000   3.840000 (  3.842717)
```

**Comment by jdantonio** [comment]
Thank you!

**Comment by pitr-ch** [comment]
@lucasallan I very appreciate your contributions but this particular change is unsafe. The @state instance variable is not properly initialized. Generally speaking I think all of our abstractions should use synchronisation layer to avoid this type of problems. It also helps greatly to solve issues across the gem if one is found. I know the benchmark shows improvement but when the number of iterations is taken into-account it's not big gain. It's also best to avoid optimisations of a method inlining kind, that's a job for the Ruby interpreter. We should prefer code clarity and algorithmic optimisations. I would suggest to revert this change.

**Comment by lucasallan** [comment]
@pitr-ch Could you clarify what you meant by `is not properly initialized`? Also, we're using `Concurrent:: AtomicReference` which already give us some benefits + performance, I don't understand why it would be a bad idea.

**Comment by pitr-ch** [comment]
Compiler or processor can reorder writes, in particular it may choose to reorder write to `@state` variable with write of the newly created `Atom` object to a variable accessible by other threads. If it does than another thread may observe `Atom` object with @state variable still uninitialized which would lead to `#value` method failing on nil error.

There is nothing bad about `Concurrent::AtomicReference`, in fact `attr_volatile_with_cas` uses it to implement CAS operations over the state field. The problematic part is that this is already abstracted pattern and it's expected to be different in future based on Ruby implementation. Some Ruby implementations may implement this directly (CAS operations over instance variables) which will allow to get rid of the indirection caused by `AtomicReference`. When this happens this implementation would not get the benefit of the abstraction provided by the layer, but it would have to be modified again to use the layer.

**Comment by lucasallan** [comment]
Fair enough. Thanks for the awesome explanation @pitr-ch - I'll revert it later today.

**Comment by pitr-ch** [comment]
Thanks @lucasallan.

**Pull request merged**